File: PandorasBox/Ahorn/entities/lasernoteblock.jl
module PandorasBoxLaserNoteBlock

using ..Ahorn, Maple

@mapdef Entity "pandorasBox/laserNoteBlock" LaserNoteBlock(x::Integer, y::Integer, pitch::Int=69, volume::Number=1.0, sound::String="game_03_deskbell_again", direction::String="Horizontal", atPlayer::Bool=false)

const directions = String["Horizontal", "Vertical"]

const placements = Ahorn.PlacementDict(
    "Laser Note Block ($direction) (Pandora's Box)" => Ahorn.EntityPlacement(
        LaserNoteBlock,
        "point",
        Dict{String, Any}(
            "direction" => direction
        )
    ) for direction in directions
)

Ahorn.editingOptions(entity::LaserNoteBlock) = Dict{String, Any}(
    "direction" => directions
)

const sprites = Dict{String, String}(
    "Horizontal" => "objects/pandorasBox/laser/noteblock/noteblock_horizontal",
    "Vertical" => "objects/pandorasBox/laser/noteblock/noteblock_vertical"
)

function Ahorn.selection(entity::LaserNoteBlock)
    x, y = Ahorn.position(entity)
    direction = get(entity, "direction", "Horizontal")
    sprite = get(sprites, direction, sprites["Horizontal"])

    return Ahorn.getSpriteRectangle(sprite, x, y, jx=0.5, jy=0.5)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::LaserNoteBlock, room::Maple.Room)
    direction = get(entity, "direction", "Horizontal")
    sprite = get(sprites, direction, sprites["Horizontal"])

    Ahorn.drawSprite(ctx, sprite, 0, 0, jx=0.5, jy=0.5)
end

end

File: PandorasBox/Loenn/entities/laser_note_block.lua
local laserNoteBlock = {}

local directions = {
    "Horizontal",
    "Vertical"
}

laserNoteBlock.name = "pandorasBox/laserNoteBlock"
laserNoteBlock.depth = 50
laserNoteBlock.fieldInformation = {
    direction = {
        options = directions,
        editable = false
    }
}
laserNoteBlock.placements = {}

for _, direction in ipairs(directions) do
    table.insert(laserNoteBlock.placements, {
        name = string.lower(direction),
        data = {
            direction = direction,
            pitch = 69,
            volume = 1.0,
            sound = "game_03_deskbell_again",
            atPlayer = false
        }
    })
end

local horizontalTexture = "objects/pandorasBox/laser/noteblock/noteblock_horizontal"
local verticalTexture = "objects/pandorasBox/laser/noteblock/noteblock_vertical"

function laserNoteBlock.texture(room, entity)
    local horizontal = string.lower(entity.direction or "Horizontal") == "horizontal"

    return horizontal and horizontalTexture or verticalTexture
end

return laserNoteBlock